File: next-lexicographical-permutation-algorithm/nextperm.cs
/* 
 * Next lexicographical permutation algorithm (C#)
 * by Project Nayuki, 2016. Public domain.
 * https://www.nayuki.io/page/next-lexicographical-permutation-algorithm
 */


public sealed class nextperm {
	
	// Computes the next lexicographical permutation of the given array
	// of integers in place, returning whether a next permutation existed.
	// (Returns false when the argument is already the last possible permutation.)
	public static bool NextPermutation(int[] array) {
		// Find non-increasing suffix
		int i = array.Length - 1;
		while (i > 0 && array[i - 1] >= array[i])
			i--;
		if (i <= 0)
			return false;
		
		// Find successor to pivot
		int j = array.Length - 1;
		while (array[j] <= array[i - 1])
			j--;
		int temp = array[i - 1];
		array[i - 1] = array[j];
		array[j] = temp;
		
		// Reverse suffix
		j = array.Length - 1;
		while (i < j) {
			temp = array[i];
			array[i] = array[j];
			array[j] = temp;
			i++;
			j--;
		}
		return true;
	}
	
}

File: next-lexicographical-permutation-algorithm/nextperm.m
% 
% Next lexicographical permutation algorithm (MATLAB)
% by Project Nayuki, 2017. Public domain.
% https://www.nayuki.io/page/next-lexicographical-permutation-algorithm
% 


function result = nextperm(arr)
% Computes and returns the next lexicographical permutation of the given vector,
% or returns [] when the argument is already the last possible permutation.
% Example: nextperm([0, 1, 0]) -> [1, 0, 0]

	% Find non-increasing suffix
	i = length(arr);
	while i > 1 && arr(i - 1) >= arr(i)
		i = i - 1;
	end
	if i <= 1
		result = [];
		return;
	end
	
	% Find successor to pivot
	result = arr;
	j = length(result);
	while result(j) <= result(i - 1)
		j = j - 1;
	end
	temp = result(i - 1);
	result(i - 1) = result(j);
	result(j) = temp;
	
	% Reverse suffix
	result(i : end) = result(end : -1 : i);
end
